src/css3/new_tokenizer.rs

```diff
@@ -131,30 +335,128 @@ mod test {
+        }
+    }
+
+    // todo: add more tests for the `<ident-token>`
+    #[test]
+    fn parse_ident_tokens() {
+        let mut is = InputStream::new();
+
+        let ident_tokens = vec![
+            ("-ident", "-ident"),
+            ("ide  nt", "ide"),
+            ("_123-ident", "_123-ident"),
+            ("_123\\ident", "_123"),
         ];
 
-        for expected_number in expected_numbers {
-            assert_eq!(tokenizer.consume_number(), expected_number);
-            tokenizer.stream.read_char(); // '#'
+        let mut tokenizer = Tokenizer::new(&mut is);
+
+        for (raw_ident, ident_tokens) in ident_tokens {
+            tokenizer
+                .stream
+                .read_from_str(raw_ident, Some(Encoding::UTF8));
+
+            assert_eq!(tokenizer.consume_ident(), ident_tokens);
         }
+    }
 
-        assert!(is.eof())
+    #[test]
+    fn parse_escaped_tokens() {
+        {
+            let mut is = InputStream::new();
+
+            let escaped_chars = vec![
+                ("\\005F ", get_unicode_char(UnicodeChar::LowLine)),
+                ("\\2A", '*'),
+                (
+                    "\\000000 ",
+                    get_unicode_char(UnicodeChar::ReplacementCharacter),
+                ),
+                (
+                    "\\FFFFFF ",
+                    get_unicode_char(UnicodeChar::ReplacementCharacter),
+                ),
+                (
+                    "\\10FFFF ",
+                    get_unicode_char(UnicodeChar::ReplacementCharacter),
+                ),
+            ];
+
+            let mut tokenizer = Tokenizer::new(&mut is);
+
+            for (raw_escaped, escaped_char) in escaped_chars {
+                tokenizer
+                    .stream
+                    .read_from_str(raw_escaped, Some(Encoding::UTF8));
+                assert_eq!(tokenizer.consume_escaped_token(), escaped_char);
+            }
+        }
+    }
+
+    #[test]
+    fn parse_urls() {
+        let mut is = InputStream::new();
+
+        let urls = vec![
+            (
+                "url(https://gosub.io/)",
+                Token::Url("https://gosub.io/".into()),
+            ),
+            ("url(  gosub.io   )", Token::Url("gosub.io".into())),
+            ("url(gosub\u{002E}io)", Token::Url("gosub.io".into())),
+            ("url(gosub\u{FFFD}io)", Token::Url("gosub�io".into())),
+            ("url(gosub\u{0000}io)", Token::BadUrl("gosub".into())),
+        ];
+
+        let mut tokenizer = Tokenizer::new(&mut is);
+
+        for (raw_url, url_token) in urls {
+            tokenizer
+                .stream
+                .read_from_str(raw_url, Some(Encoding::UTF8));
+            assert_eq!(tokenizer.consume_ident_like_seq(), url_token);
+        }
+    }
+
+    #[test]
+    fn parse_function_tokens() {
+        let mut is = InputStream::new();
+
+        let functions = vec![
+            ("url(\"", Token::Function("url".into())),
```

Comment:
Should the other cases mentioned in 4.3.4 be added here such as `url( \"`, `url(\'` and `url( \')`?